client/internal/routemanager/client.go

```diff
@@ -0,0 +1,258 @@
+			return
+		case <-c.statusRecorder.GetPeerStateChangeNotifier(peer):
+			peerStateUpdate <- struct{}{}
+			log.Debugf("triggered state update for Peer %s", peer)
+		}
+	}
+}
+
+func (c *clientNetwork) startPeersStatusChangeWatcher() {
+	for _, r := range c.routes {
+		_, found := c.routePeersNotifiers[r.Peer]
+		if !found {
+			c.routePeersNotifiers[r.Peer] = make(chan struct{})
+			go c.watchPeerStatusChanges(c.ctx, r.Peer, c.peerStateUpdate, c.routePeersNotifiers[r.Peer])
+		}
+	}
+}
+
+func (c *clientNetwork) removeRouteFromPeerAndSystem() error {
+	if c.chosenRoute != nil {
+		err := c.wgInterface.RemoveAllowedIP(c.chosenRoute.Peer, c.network.String())
+		if err != nil {
+			return fmt.Errorf("couldn't remove allowed IP %s removed for peer %s, err: %v",
+				c.network, c.chosenRoute.Peer, err)
+		}
+		err = removeFromRouteTable(c.network)
+		if err != nil {
+			return fmt.Errorf("couldn't remove route %s from system, err: %v",
+				c.network, err)
+		}
+	}
+	return nil
+}
+
+func (c *clientNetwork) recalculateRouteAndUpdatePeerAndSystem() error {
+
+	var err error
+
+	routerPeerStatuses := c.getRouterPeerStatuses()
+
+	chosen := c.getBestRouteFromStatuses(routerPeerStatuses)
+	if chosen == "" {
+		err = c.removeRouteFromPeerAndSystem()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if c.chosenRoute != nil && c.chosenRoute.ID == chosen {
+		if c.chosenRoute.IsEqual(c.routes[chosen]) {
+			return nil
+		}
+	}
+
+	if c.chosenRoute != nil {
+		err = c.wgInterface.RemoveAllowedIP(c.chosenRoute.Peer, c.network.String())
+		if err != nil {
+			return fmt.Errorf("couldn't remove allowed IP %s removed from previously chosed peer %s, err: %v",
+				c.network, c.chosenRoute.Peer, err)
+		}
+	} else {
+		err = addToRouteTable(c.network, c.wgInterface.GetAddress().IP.String())
+		if err != nil {
+			return fmt.Errorf("route %s couldn't be added for peer %s, err: %v",
+				c.chosenRoute.Network.String(), c.wgInterface.GetAddress().IP.String(), err)
+		}
+	}
+
+	c.chosenRoute = c.routes[chosen]
+	err = c.wgInterface.AddAllowedIP(c.chosenRoute.Peer, c.network.String())
+	if err != nil {
+		log.Errorf("couldn't add allowed IP %s added for peer %s, err: %v",
+			c.network, c.chosenRoute.Peer, err)
+	}
+
+	return nil
+}
+
+func (c *clientNetwork) handleUpdate(update routesUpdate) {
+	if update.updateSerial < c.updateSerial {
+		log.Warnf("received a routes update with smaller serial number, ignoring it")
+		return
+	}
+
+	updateMap := make(map[string]*route.Route)
+
+	for _, r := range update.routes {
+		updateMap[r.ID] = r
+	}
+
+	for id, r := range c.routes {
+		_, found := updateMap[id]
+		if !found {
+			close(c.routePeersNotifiers[r.Peer])
+		}
+	}
+
+	c.routes = updateMap
+	c.updateSerial = update.updateSerial
+}
+
+// stateAndUpdateWatcher is the main point of reacting on client network routing events.
+// All the processing related to the client network should be done here. Thread-safe.
+func (c *clientNetwork) stateAndUpdateWatcher() {
```

Comment:
stateAndUpdateWatcher is a typo? 
startAndUpdateWatcher?

Reply:
was state, but I renamed it to peersStateAndUpdateWatcher